### PR TITLE
sycl - fix vec memory copy

### LINKED ISSUE #1545: SYCL Regression on Sunspot
So running tests on Sunspot, I get the following error message running `t106-vector /gpu/self/ref`:
```
Exception caught at File: queue.hpp | Function: copy | Line: 672 | Column: 12
terminate called after throwing an instance of 'sycl::_V1::runtime_error'
  what():  NULL pointer argument in memory copy operation. -30 (PI_ERROR_INVALID_VALUE)
```

The last two lines are the same as other errors we've seen. Here's the full backtrace:
```
#0  0x00001555545ebc6b in raise () from /lib64/libc.so.6
#1  0x00001555545ed305 in abort () from /lib64/libc.so.6
#2  0x0000155552669675 in ?? () from /usr/lib64/libstdc++.so.6
#3  0x0000155552674d0c in ?? () from /usr/lib64/libstdc++.so.6
#4  0x0000155552674d77 in std::terminate() () from /usr/lib64/libstdc++.so.6
#5  0x0000155552674fd8 in __cxa_throw () from /usr/lib64/libstdc++.so.6
#6  0x000015555227f4f3 in sycl::_V1::detail::queue_impl::memcpy(std::shared_ptr<sycl::_V1::detail::queue_impl> const&, void*, void const*, unsigned long, std::vector<sycl::_V1::event, std::allocator<sycl::_V1::event> > const&, sycl::_V1::detail::code_location const&) () from /opt/aurora/23.275.2/CNDA/oneapi/compiler/eng-20231130/lib/libsycl.so.7
#7  0x0000155552330946 in sycl::_V1::queue::memcpy(void*, void const*, unsigned long, sycl::_V1::event, sycl::_V1::detail::code_location const&) () from /opt/aurora/23.275.2/CNDA/oneapi/compiler/eng-20231130/lib/libsycl.so.7
#8  0x000015555495fe6c in sycl::_V1::queue::copy<double> (this=0x42cb30, Src=Src@entry=0xff002aaaaaa80000, Dest=0x0, Count=<optimized out>, DepEvent=..., CodeLoc=...)    at /opt/aurora/23.275.2/CNDA/oneapi/compiler/eng-20231130/bin/compiler/../../include/sycl/queue.hpp:672
#9  0x0000155554976cc7 in CeedVectorSetArrayDevice_Sycl (vec=<optimized out>, copy_mode=<optimized out>, array=0xff002aaaaaa80000) at /home/jrwrigh/software/libCEED/backends/sycl-ref/ceed-sycl-vector.sycl.cpp:218
#10 CeedVectorSetArray_Sycl (vec=<optimized out>, mem_type=<optimized out>, copy_mode=<optimized out>, array=0xff002aaaaaa80000) at /home/jrwrigh/software/libCEED/backends/sycl-ref/ceed-sycl-vector.sycl.cpp:263
#11 0x000015555491ead4 in CeedVectorSetArray (vec=0x247acc0, mem_type=(unknown: 0xfffef360), mem_type@entry=CEED_MEM_DEVICE, copy_mode=copy_mode@entry=CEED_COPY_VALUES, array=0x1555545ebc6b <raise+269>)    at /home/jrwrigh/software/libCEED/interface/ceed-vector.c:267
#12 0x0000000000400b9b in main (argc=<optimized out>, argv=<optimized out>) at /home/jrwrigh/software/libCEED/tests/t106-vector.c:30
```

CC: @jeremylt

### PR DESCRIPTION
**Pull request by jeremylt**
Closes #1545 

Oops

### CODE REVIEW

**Comment by jrwrigh** [comment]
Lol. I was about to write on the issue that `impl->d_vector` was `NULL`, but `d_vector_owned` was `non-NULL`. I'll verify the fix real quick.

**Comment by jeremylt** [comment]
Note. CI blocked by 1544

**Comment by jrwrigh** [comment]
So that does fix the issue in #1545, but I'm now running into a different issue...

**Review by jrwrigh — APPROVED** [review]

**Comment by jeremylt** [comment]
If you find more fixes before the morning. Feel free to push them here

**Comment by jrwrigh** [comment]
The only other test failure I'm seeing is for `t506-operator`, but it's failing on correctness, not erroring out. And it's only for `/gpu/self/ref`:
```
    ok 4 - , /cpu/self/opt/blocked
      ---
      args: build/t506-operator /cpu/self/opt/blocked
      ...
    not ok 5 - , /gpu/sycl/ref
      ---
      args: build/t506-operator /gpu/sycl/ref
      num_failures: 1
      failure_0: stdout
        message: stdout
        output: |
          Computed Area: 1.254524 != True Area: 1.0
      ...
    ok 6 - , /gpu/sycl/shared
      ---
      args: build/t506-operator /gpu/sycl/shared
      ...
    ok 7 - , /gpu/sycl/gen
      ---
      args: build/t506-operator /gpu/sycl/gen
      ...
not ok 1 -
```

The `t506` header mentions reuse of QFunctions, so perhaps missing a zeroing out of a vector?

**Comment by jeremylt** [comment]
Its supposed to check that using the same QF between two operators isn't a problem. One way this came up was OCCA hardwiring the size of Q into the JiT

**Comment by jeremylt** [comment]
I would guess this is an issue somewhere in `ceed-sycl-ref-basis.sycl.cpp`

File: backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp
Comment (jrwrigh) [inline backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp:1]: Unfortunately doesn't look like SYCL likes this for some reason. Seeing the same thing on Sunspot that the SYCL build job is reporting.

Comment (jeremylt) [reply backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp:1]: Why?!? That's silly.

Comment (jrwrigh) [reply backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp:1]: :shrug: I am not wise in the ways of SYCL/C++ templates, so I don't know.

File: backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp [hunk 1]
```diff
@@ -468,10 +460,18 @@ static int CeedBasisApplyNonTensor_Sycl(CeedBasis basis, const CeedInt num_elem,
   // Apply basis operation
   switch (eval_mode) {
     case CEED_EVAL_INTERP: {
-      CeedCallBackend(CeedBasisApplyNonTensorInterp_Sycl(data->sycl_queue, num_elem, is_transpose, impl, d_u, d_v));
+      if (is_transpose) {
+        CeedCallBackend(CeedBasisApplyInterp_Sycl<true>(data->sycl_queue, *impl->sycl_module, num_elem, impl, d_u, d_v));
+      } else {
+        CeedCallBackend(CeedBasisApplyInterp_Sycl<false>(data->sycl_queue, *impl->sycl_module, num_elem, impl, d_u, d_v));
```
Comment (jrwrigh) [inline backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp:466]: Doesn't like `true` or `false` either, different complaint though:
```
/home/jrwrigh/software/libCEED/backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp:464:82: error: no member named 'sycl_module' in 'CeedBasisNonTensor_Sycl'
  464 |         CeedCallBackend(CeedBasisApplyInterp_Sycl<true>(data->sycl_queue, *impl->sycl_module, num_elem, impl, d_u, d_v));
      |                                                                            ~~~~  ^
./include/ceed/backend.h:214:17: note: expanded from macro 'CeedCallBackend'
  214 |     int ierr_ = __VA_ARGS__;                                                     \
      |                 ^~~~~~~~~~~
```

Comment (jrwrigh) [reply backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp:466]: Oh, actually just looks like you're calling a different function now.

Comment (jeremylt) [reply backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp:466]: That argument should have never actually been an int

Comment (jeremylt) [reply backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp:466]: ugh

**Comment by jrwrigh** [comment]
On the bright side, (reverting the code failing to compile), all the tests pass now. Looks the same as it did last night, so it might've been some weird build issue on my side.

File: backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp (hunk 1 above)
Comment (jeremylt) [reply backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp]: Should be good now

**Review by jrwrigh — APPROVED** [review]
Confirmed that the current branch state passes all core tests. I'm getting all kinds of fun failures for fluids, but I think we might as well go ahead and merge this while I wrestle with those.

**Pull request merged**